src/frontend/radeco_containers.rs

```diff
@@ -110,47 +161,272 @@ pub enum FunctionType {
     Import(u16),
 }
 
-pub const UNKNOWN_MODULE: u16 = 0xFFFF;
-
-#[derive(Debug)]
+#[derive(Debug, Clone, Default)]
 /// Container to store information about identified function.
 /// Used as a basic unit in intra-functional analysis.
 pub struct RadecoFunction {
-    /// Identified variable bindings for `RadecoFunction`
-    bindings: RadecoBindings<Binding<NodeIndex>>,
-    /// Current module being analyzed, i.e., the index of the module that returned this
-    /// RadecoFunction instance. Note that this is not necessarily
-    /// the module where the function is defined, in case of an import.
-    current_module: u16,
-    /// Represents the type of function
-    ftype: FunctionType,
+    // Identified variable bindings for `RadecoFunction`
+    // bindings: RadecoBindings<Binding<NodeIndex>>,
+    // Represents the type of function
+    // ftype: FunctionType,
     /// Raw instruction information for the current function
     pub instructions: Vec<LOpInfo>,
     /// Is current function known to be recursive
-    is_recursive: bool,
+    is_recursive: Option<bool>,
     /// Human readable name for the function. Taken either from
     /// the symbol table or assigned based on offset.
     pub name: Cow<'static, str>,
     /// Start address of the function
     pub offset: u64,
-    /// Calls to the function
-    refs: Vec<CallRefs>,
+    // TODO: Calls to the function
+    // refs: Vec<CallRefs>,
     /// Size of the function in bytes
     size: u64,
     /// Constructed SSA for the function
-    ssa: SSAStorage,
-    /// Calls from the current function to other functions
-    xrefs: Vec<CallRefs>,
+    ssa: SSAStorage, /* TODO: Calls from the current function to other functions
+                      * xrefs: Vec<CallRefs>, */
 }
 
-#[derive(Debug)]
-/// Default project loader that uses r2 and r2pipe
-/// Implements the `ProjectLoader` trait
-pub struct R2ProjectLoader { }
+#[derive(Default)]
+pub struct ProjectLoader<'a> {
+    load_libs: bool,
+    path: Cow<'static, str>,
+    load_library_path: Option<Cow<'static, str>>,
+    filter_modules: Option<fn(&RadecoModule) -> bool>,
+    source: Option<Rc<Source>>,
+    mloader: Option<ModuleLoader<'a>>,
+}
 
-#[derive(Debug)]
-/// Load a RadecoProject from formatted JSON files
-pub struct JSONProjectLoader { }
+impl<'a> ProjectLoader<'a> {
+    /// Enable loading of libraries
+    pub fn load_libs(mut self) -> ProjectLoader<'a> {
+        self.load_libs = true;
+        self
+    }
+
+    /// Set path to load from
+    pub fn path<T: AsRef<str>>(mut self, path: T) -> ProjectLoader<'a> {
+        self.path = Cow::from(path.as_ref().to_owned());
+        self
+    }
+
+    /// Setup and configure `ModuleLoader` to use
+    pub fn module_loader<'b: 'a>(mut self, mloader: ModuleLoader<'b>) -> ProjectLoader<'a> {
+        self.mloader = Some(mloader);
+        self
+    }
+
+    /// Set the source to use for loading. This is propagated to every `ModuleLoader`
```

Comment:
Why do we need to replicate the entire Source for each ModuleLoader? Naive doubt, but doesnt it increase redundancy?

Reply:
This allows all the modules loaders to work independently to load themselves up (using source). It also allows us to load different modules from different sources.

For example, you could load the main binary and perform analysis using r2 (-source), but for the libraries (which don't change regardless of which binary they're used in) we could analyze them once, cache the analysis, and reload using the cached source.

From memory usage perspective, this isn't heavy, as all we're doing is storing a reference to the original source (as opposed to cloning it).